<a href="https://colab.research.google.com/github/FERNANDOHENZ/CODENATION/blob/master/Desafio_M%C3%B3dulo_2_IGTI_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Importando as Bibliotecas e o Dataset**


In [0]:
#importando as bibliotecas
import pandas as pd #biblioteca utilizada para o tratamento de dados via dataframes 
import numpy as np #biblioteca utilizada para o tratamento de valores numéricos (vetores e matrizes)
import matplotlib.pyplot as plt #biblioteca utilizada para construir os gráficos




In [135]:
from google.colab import files
uploaded = files.upload()

Saving winequality-red.csv to winequality-red (1).csv


In [138]:
#realizando a leitura do dataset, aqui temos que colocar para ler o separador ' ; ' para ler o dtaset
ds_wine = pd.read_csv('winequality-red.csv', sep=';')
ds_wine.head()

fixed acidity  volatile acidity  citric acid  ...  sulphates  alcohol  quality
0            7.4              0.70         0.00  ...       0.56      9.4        5
1            7.8              0.88         0.00  ...       0.68      9.8        5
2            7.8              0.76         0.04  ...       0.65      9.8        5
3           11.2              0.28         0.56  ...       0.58      9.8        6
4            7.4              0.70         0.00  ...       0.56      9.4        5

[5 rows x 12 columns]

# **Resolução Pergunta 1**

In [139]:
ds_wine.shape

(1599, 12)

**Pergunta 1 :** Quantas instâncias e atributos existem respectivamente?

Resposta: 1599 Instâncias e 12 atributos

# **Resolução Perguntas 2 e 3**

In [142]:
ds_wine.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1599 entries, 0 to 1598
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         1599 non-null   float64
 1   volatile acidity      1599 non-null   float64
 2   citric acid           1599 non-null   float64
 3   residual sugar        1599 non-null   float64
 4   chlorides             1599 non-null   float64
 5   free sulfur dioxide   1599 non-null   float64
 6   total sulfur dioxide  1599 non-null   float64
 7   density               1599 non-null   float64
 8   pH                    1599 non-null   float64
 9   sulphates             1599 non-null   float64
 10  alcohol               1599 non-null   float64
 11  quality               1599 non-null   int64  
dtypes: float64(11), int64(1)
memory usage: 150.0 KB


In [143]:
ds_wine.dtypes.nunique()

2

**Pergunta 2**: Quantos tipos diferentes de dados existem no dataset do desafio?

Resposta : 2 (Float e integer)

**Pergunta 3:** Qual é a proporção (em %) de valores nulos existente na coluna "residual sugar" (resíduo de açucar)?

Resposta: 0% (é possivel ver já com a informação de info()

# **Resolução Pergunta 4**

In [144]:
#Descobrindo o desvio padrão das váriaveis
ds_wine.describe()

fixed acidity  volatile acidity  ...      alcohol      quality
count    1599.000000       1599.000000  ...  1599.000000  1599.000000
mean        8.319637          0.527821  ...    10.422983     5.636023
std         1.741096          0.179060  ...     1.065668     0.807569
min         4.600000          0.120000  ...     8.400000     3.000000
25%         7.100000          0.390000  ...     9.500000     5.000000
50%         7.900000          0.520000  ...    10.200000     6.000000
75%         9.200000          0.640000  ...    11.100000     6.000000
max        15.900000          1.580000  ...    14.900000     8.000000

[8 rows x 12 columns]

In [145]:
#Forma alternativa para ver o desvio padrão
std_fixed_acidity = ds_wine['fixed acidity'].std()
print(std_fixed_acidity)

1.7410963181277006


In [146]:
# outra forma para ver o desvio padrão
ds_wine['fixed acidity'].std()

1.7410963181277006

**Pergunta 4** - Qual é o desvio padrão para a variável "fixed acidity" ?

Resposta 1.74

# **Resolução Pergunta 5**

Qual é a mediana para a variável "residual sugar"?

In [147]:
# Descobrindo a medianda, ela corresponde ao valor de 50% obtido no info anteriormente que seria igual a 2.2
# Vamos apresentar outra forma de resolver
ds_wine['residual sugar'].median()



2.2

Resposta : a Mediana é igual a 2.2

# **Resolução Pergunta 6**

Sobre o coeficiente de correlação de Pearson entre as variáveis "fixed acidity" e "pH" é CORRETO afirmar:


1. Existe uma correlação positiva. Quanto maior o "fixed acidity, menor é o "pH".

2. Existe uma correlação negativa. Quanto maior o "fixed acidity, menor é o "pH".

3. Existe uma correlação negativa. Quanto maior o "fixed acidity, maior é o "pH".

4. Existe uma correlação negativa. Quanto menor o "fixed acidity, menor é o "pH".





In [148]:
ds_wine[['fixed acidity', 'pH']].corr()

fixed acidity        pH
fixed acidity       1.000000 -0.682978
pH                 -0.682978  1.000000

Resposta:
Existe uma correlação negativa, quanto mais alto fixed acidity menor é o pH. Alternativa 1

# **Resolução Pergunta 7**

Mesmo sendo reduzido, é CORRETO dizer sobre o coeficiente de correlação de Pearson entre as variáveis "quality" e "alcohol":

1. Quanto maior o teor alcoolico (alcohol), maior é a qualidade do vinho (quality).

2. Não é possível dizer nada sobre a correlação entre as vairáveis.

3. Para analisar correlação é necessário que o coeficiente seja superior a 1.

4. Quanto maior o teor alcoolico (alcohol), menor é a qualidade do vinho (quality).

In [149]:
ds_wine[['quality', 'alcohol']].corr()

quality   alcohol
quality  1.000000  0.476166
alcohol  0.476166  1.000000

In [150]:
ds_wine[['alcohol', 'quality']].corr()

alcohol   quality
alcohol  1.000000  0.476166
quality  0.476166  1.000000

Resposta: Qunato maior o teor de alcool maior a qualidade do vinho

# **Resolução da Pergunta  8**

Quantas instâncias existem para a qualidade do vinho igual a 5?

Grupo de escolhas da pergunta

53

638

681

345

In [151]:
ds_wine['quality'].value_counts()

5    681
6    638
7    199
4     53
8     18
3     10
Name: quality, dtype: int64

In [152]:
len(ds_wine[ds_wine['quality']==5])

681

Resposta 681

# **Resolução da Pergunta 9**

Aplique o modelo de normalização MinMaxScaler com os valores "default" para os dados de entrada. Qual é o menor valor para a variável "fixed acidity"?

Grupo de escolhas da pergunta

0

15.2

1

4,6

In [0]:
# Separando as váriaveis independentes e dependentes

X = ds_wine.drop(['quality'], axis = 1)
y = ds_wine['quality']

In [0]:
# Efetuandi a Normalização dos dados
# Importando a Bibiblioteca MinMaxScaler
from sklearn.preprocessing import MinMaxScaler


In [0]:
# Instanciando a função MinMaxScaler
scaler = MinMaxScaler()


In [0]:
#Aplicando a função MinMaxScaler
X = scaler.fit_transform(X)


In [157]:
pd.DataFrame( X, columns = ds_wine.columns[:-1])['fixed acidity'].min()


0.0

Resultado = 0

**# Resolução da Questão 10**
Após dividir o dataset utilizando o método train_test_split(entradas_normalizadas, saida, test_size=0.30, random_state=1). Qual foi a acurária para o KNN?

Grupo de escolhas da pergunta

0,89

1,2

0,56

0,15



In [0]:
# Dividindo a base de treino e teste
from sklearn.model_selection import train_test_split


In [0]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = 1)

In [0]:
# Importanto o Algoritmo KNN
from sklearn.neighbors import KNeighborsClassifier

In [0]:
# Instanciando o Algoritmo
clf_KNN = KNeighborsClassifier(n_neighbors= 5)


In [162]:
# Treinando o modelo
clf_KNN.fit(X_train, y_train)


KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
                     metric_params=None, n_jobs=None, n_neighbors=5, p=2,
                     weights='uniform')

In [163]:
#Medindo a acuracia
clf_KNN.score(X_test, y_test)

0.5645833333333333

Resposta Acurácia do KNN = 0.56

# **Resolução da Pergunta 11**
Aplicando o algoritmo de árvore de decisão com os parâmetros definidos no enunciado do desafio e a divisão de treinamento e teste como mostrado, anteriormente. Qual é o valor da acurácia?

Grupo de escolhas da pergunta

0.80

-0.80

0.58

-0.58

In [0]:
# Importando o algoritimo da árvore de decisão
from sklearn.tree import DecisionTreeClassifier

In [0]:
# Instanciando o algoritmo
clf_arvore = DecisionTreeClassifier()


In [166]:
#Efetuando o treinamento do modelo
clf_arvore.fit(X_train, y_train)

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='gini',
                       max_depth=None, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=None, splitter='best')

In [167]:
# Qual o score ou acurácia do modelo 
clf_arvore.score(X_test, y_test)

0.5875

Resposta: A medida que treinamos o modelo o valor muda um pouco, chegamos a uma curácia de 0.59 portanto acredito que a resposta seja 0.58

# **Resolução Pergunta 12**

Aplicando o algoritmo de floresta randomica com os mesmos parâmetros apresentados no enunciado e mesmos valores treinamento e teste, qual é o valor da acurácia?

Grupo de escolhas da pergunta

0,68

0,30

1,00

0,40

In [0]:
#Importando o algoritmo da floresta Randômica

from sklearn.ensemble import RandomForestClassifier


In [0]:
# Insntanciando o algoritmo
clf_floresta = RandomForestClassifier(max_depth = 10, random_state = 1)


In [170]:
# Efetuando o treinamento do modelo
clf_floresta.fit(X_train, y_train)


RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=10, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=100,
                       n_jobs=None, oob_score=False, random_state=1, verbose=0,
                       warm_start=False)

In [171]:
print('Acurácia: {}'.format(clf_floresta.score(X_test, y_test)))

Acurácia: 0.6791666666666667


Resposta A Acurácia deste modelo chegou em 0.67 portanto alternativa 0..68

# **Pergunta 13**

Aplicando o algoritmo SVM para a classificação da qualidade dos vinhos com os mesmos valores anteriores para treinamento e teste, qual é o valor de acurácia obtido?

  0,36   

  0,59 

  0,72 

  0,15 

In [0]:
# Importando o Algoritmo SVM - Support Vector Machine(SVC)
from sklearn.svm import SVC

In [0]:
# Instanciando o algoritmo
clf_svm = SVC(gamma = 'auto', kernel = 'rbf')

In [174]:
# Efetuando o treinamento do modelo
clf_svm.fit(X_train, y_train)

SVC(C=1.0, break_ties=False, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma='auto', kernel='rbf',
    max_iter=-1, probability=False, random_state=None, shrinking=True,
    tol=0.001, verbose=False)

In [0]:
# Fazendo as previsões
prev_svm = clf_svm.predict(X_test)


In [176]:
# Apurando a acurácia
clf_svm.score(X_test, y_test)

0.5854166666666667

Resposta : 
A Acurácia desse modelo foi igual a 0.58 portanto a resposta é 0.59

# **Resolução Pergunta 14**

Apicando a rede MLP como mostrado no enunciado e com os mesmos dados de treinamento e teste utilizados, qual é a acurácia do modelo?

  0,25 

  0,30 

  0,90 

  0,60 
  

In [0]:
# Importando o algoritmo MLP
from sklearn.neural_network import MLPClassifier

In [0]:
# Instanciando o algoritmo
clf_mlp = MLPClassifier(alpha=1e-5, hidden_layer_sizes=(5,5), random_state=1)


In [179]:
# Efetuando o treinamento do modelo
clf_mlp.fit(X_train, y_train)


/usr/local/lib/python3.6/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:571: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


MLPClassifier(activation='relu', alpha=1e-05, batch_size='auto', beta_1=0.9,
              beta_2=0.999, early_stopping=False, epsilon=1e-08,
              hidden_layer_sizes=(5, 5), learning_rate='constant',
              learning_rate_init=0.001, max_fun=15000, max_iter=200,
              momentum=0.9, n_iter_no_change=10, nesterovs_momentum=True,
              power_t=0.5, random_state=1, shuffle=True, solver='adam',
              tol=0.0001, validation_fraction=0.1, verbose=False,
              warm_start=False)

In [180]:
# Determinando a acurácia
clf_mlp.score(X_test, y_test)

0.6

Resposta: Aqui temos uma acurácia de 0.6 sendo esta nossa resposta

# **Pergunta 15**


Modifique o dataset para um sistema binário que considere vinhos com nota maior do que 5 (6, 7, 8) sejam bons e menores ou igual a 5 ruins (5, 4, 3). Aplique o modelo da floresta randômica. Qual a acurácia do modelo para esse novo dataset?

  0,60 

  0,50 

  0,68 

  0,80 


In [0]:
y_train = y_train > 5  # Definindo os valores acima de 5 para o conjunto treino
y_train = y_train.astype(int)# Alterando para someto números inteiros


In [0]:
y_test = y_test > 5 # Definindo os valores acima de 5 para o conjunto teste 
y_test = y_test.astype(int) # Alterando para someto números inteiros


In [0]:
# Importando o algoritmo de floresta Randômica (Random Forest)
from sklearn.ensemble import RandomForestClassifier

In [0]:
# Instanciando o algoritmo
clf_floresta = RandomForestClassifier(max_depth=10, random_state = 1)


In [185]:
# Efetuando o treinamento do modelo
clf_floresta.fit(X_train, y_train)


RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=10, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=100,
                       n_jobs=None, oob_score=False, random_state=1, verbose=0,
                       warm_start=False)

In [186]:
clf_floresta.score(X_test, y_test)

0.8041666666666667